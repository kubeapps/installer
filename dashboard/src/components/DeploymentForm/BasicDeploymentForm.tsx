import * as React from "react";

class BasicDeploymentForm extends React.Component {
  public render() {
    return <div>Basic Form!</div>;
  }
}

export default BasicDeploymentForm;
