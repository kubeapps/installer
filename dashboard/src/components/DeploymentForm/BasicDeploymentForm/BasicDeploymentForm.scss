.description {
  color: #5f6369;
  font-size: 0.9em;
}

.react-switch {
  vertical-align: middle;
  margin-left: 10px;
}

.disk_size_input {
  width: 75%;
}
