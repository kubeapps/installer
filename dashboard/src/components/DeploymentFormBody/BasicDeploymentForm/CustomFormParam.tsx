import React, { useMemo } from "react";
import { IBasicFormParam } from "shared/types";
import { CustomComponent } from "../../../RemoteComponent";

export interface ICustomParamProps {
  param: IBasicFormParam;
  handleBasicFormParamChange: (
    p: IBasicFormParam,
  ) => (e: React.FormEvent<HTMLInputElement>) => void;
}

export default function CustomFormComponentLoader({
  param,
  handleBasicFormParamChange,
}: ICustomParamProps) {
  // Fetches the custom-component bundle served by the dashboard nginx
  const url = `${window.location.origin}/custom_components.js`;

  return useMemo(
    () => (
      <CustomComponent
        url={url}
        param={param}
        handleBasicFormParamChange={handleBasicFormParamChange}
      />
    ),
    [handleBasicFormParamChange, param, url],
  );
}
