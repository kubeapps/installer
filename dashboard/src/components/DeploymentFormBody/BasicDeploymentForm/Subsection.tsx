import * as React from "react";
import { setValue } from "../../../shared/schema";
import { IBasicFormParam } from "../../../shared/types";
import { getValueFromEvent } from "../../../shared/utils";

export interface ISubsectionProps {
  label: string;
  param: IBasicFormParam;
  name: string;
  handleValuesChange: (value: string) => void;
  renderParam: (
    name: string,
    param: IBasicFormParam,
    id: string,
    handleBasicFormParamChange: (
      name: string,
      p: IBasicFormParam,
    ) => (e: React.FormEvent<HTMLInputElement>) => void,
  ) => JSX.Element | null;
  appValues: string;
}

class Subsection extends React.Component<ISubsectionProps> {
  public render() {
    const { label, param } = this.props;
    return (
      <div className="subsection margin-v-normal">
        <div className="margin-v-normal">
          {label}
          {param.description && (
            <>
              <br />
              <span className="description">{param.description}</span>
            </>
          )}
        </div>
        {param.children &&
          Object.keys(param.children).map((paramName, i) => {
            return this.props.renderParam(
              paramName,
              param.children![paramName],
              `${paramName}-${i}`,
              this.handleChildrenParamChange,
            );
          })}
      </div>
    );
  }

  private handleChildrenParamChange = (name: string, param: IBasicFormParam) => {
    return (e: React.FormEvent<HTMLInputElement>) => {
      const value = getValueFromEvent(e);
      this.props.param.children![name] = { ...this.props.param.children![name], value };
      this.props.handleValuesChange(setValue(this.props.appValues, param.path, value));
    };
  };
}

export default Subsection;
