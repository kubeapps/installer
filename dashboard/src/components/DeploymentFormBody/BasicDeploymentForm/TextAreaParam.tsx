import * as React from "react";
import { IBasicFormParam } from "shared/types";

export interface IStringParamProps {
  id: string;
  label: string;
  param: IBasicFormParam;
  handleBasicFormParamChange: (
    p: IBasicFormParam,
  ) => (e: React.FormEvent<HTMLTextAreaElement>) => void;
}

class TextAreaParam extends React.Component<IStringParamProps> {
  public render() {
    const { id, param, label } = this.props;
    return (
      <div>
        <label htmlFor={id}>
          <div className="row">
            <div className="col-3 block">
              <div className="centered">{label}</div>
            </div>
            <div className="col-9 margin-t-small">
              <textarea
                id={id}
                onChange={this.props.handleBasicFormParamChange(param)}
                value={param.value === undefined ? "" : param.value}
              />
              {param.description && <span className="description">{param.description}</span>}
            </div>
          </div>
        </label>
      </div>
    );
  }
}

export default TextAreaParam;
