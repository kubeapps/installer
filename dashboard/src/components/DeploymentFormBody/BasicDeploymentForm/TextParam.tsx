import * as React from "react";
import { IBasicFormParam } from "shared/types";

export interface IStringParamProps {
  id: string;
  name: string;
  label: string;
  inputType?: string;
  param: IBasicFormParam;
  handleBasicFormParamChange: (
    name: string,
    p: IBasicFormParam,
  ) => (e: React.FormEvent<HTMLInputElement>) => void;
}

class TextParam extends React.Component<IStringParamProps> {
  public render() {
    const { id, name, param, label, inputType } = this.props;
    return (
      <div>
        <label htmlFor={id}>
          <div className="row">
            <div className="col-3 block">
              <div className="centered">{label}</div>
            </div>
            <div className="col-9 margin-t-small">
              <input
                id={id}
                onChange={this.props.handleBasicFormParamChange(name, param)}
                value={param.value === undefined ? "" : param.value}
                type={inputType ? inputType : "text"}
              />
              {param.description && <span className="description">{param.description}</span>}
            </div>
          </div>
        </label>
      </div>
    );
  }
}

export default TextParam;
