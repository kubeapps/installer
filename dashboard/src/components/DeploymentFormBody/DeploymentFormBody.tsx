import { RouterAction } from "connected-react-router";
import * as React from "react";
import { Tab, TabList, TabPanel, Tabs } from "react-tabs";

import { retrieveBasicFormParams, setValue } from "../../shared/schema";
import { IBasicFormParam, IChartState } from "../../shared/types";
import { getValueFromEvent } from "../../shared/utils";
import { ErrorSelector } from "../ErrorAlert";
import Hint from "../Hint";
import LoadingWrapper from "../LoadingWrapper";
import AdvancedDeploymentForm from "./AdvancedDeploymentForm";
import BasicDeploymentForm from "./BasicDeploymentForm";

import "react-tabs/style/react-tabs.css";
import "./Tabs.css";

export interface IDeploymentFormBodyProps {
  chartID: string;
  chartVersion: string;
  originalValues?: string;
  namespace: string;
  releaseName?: string;
  selected: IChartState["selected"];
  appValues: string;
  valuesModified: boolean;
  push: (location: string) => RouterAction;
  goBack?: () => RouterAction;
  fetchChartVersions: (id: string) => void;
  getChartVersion: (id: string, chartVersion: string) => void;
  setValues: (values: string) => void;
  setValuesModified: () => void;
  enableBasicForm: boolean;
}

export interface IDeploymentFormBodyState {
  basicFormParameters: { [key: string]: IBasicFormParam };
}

class DeploymentFormBody extends React.Component<
  IDeploymentFormBodyProps,
  IDeploymentFormBodyState
> {
  public state: IDeploymentFormBodyState = {
    basicFormParameters: {},
  };

  public componentDidMount() {
    const { chartID, fetchChartVersions, getChartVersion, chartVersion } = this.props;
    fetchChartVersions(chartID);
    getChartVersion(chartID, chartVersion);
  }

  public componentWillReceiveProps = (nextProps: IDeploymentFormBodyProps) => {
    const { chartID, chartVersion, getChartVersion } = this.props;

    if (chartVersion !== nextProps.chartVersion) {
      // New version detected
      getChartVersion(chartID, nextProps.chartVersion);
      return;
    }

    if (nextProps.selected !== this.props.selected) {
      // The values or the schema has changed
      let values = "";
      if (!this.props.valuesModified) {
        // If the version is the current one, reuse original params
        // (this only applies to the upgrade form that has originalValues defined)
        if (
          nextProps.selected.version &&
          nextProps.selected.version.attributes.version === this.props.chartVersion &&
          this.props.originalValues
        ) {
          values = this.props.originalValues || "";
        } else {
          // In other case, use the default values for the selected version
          values = nextProps.selected.values || "";
        }
        this.props.setValues(values);
      } else {
        // If the user has modified the values, use the ones defined
        values = this.props.appValues;
      }
      if (nextProps.selected.schema) {
        this.setState({
          basicFormParameters: retrieveBasicFormParams(values, nextProps.selected.schema),
        });
      }
      return;
    }
  };

  public render() {
    const { selected, chartID, chartVersion, goBack, appValues } = this.props;
    const { version, versions } = selected;
    if (selected.error) {
      return (
        <ErrorSelector error={selected.error} resource={`Chart "${chartID}" (${chartVersion})`} />
      );
    }
    if (!version || !versions.length) {
      return <LoadingWrapper />;
    }
    return (
      <div>
        <div>
          <label htmlFor="chartVersion">Version</label>
          <select
            id="chartVersion"
            onChange={this.handleChartVersionChange}
            value={version.attributes.version}
            required={true}
          >
            {versions.map(v => (
              <option key={v.id} value={v.attributes.version}>
                {v.attributes.version}{" "}
                {this.props.releaseName && v.attributes.version === this.props.chartVersion
                  ? "(current)"
                  : ""}
              </option>
            ))}
          </select>
        </div>
        {this.shouldRenderBasicForm() ? (
          this.renderTabs()
        ) : (
          <AdvancedDeploymentForm
            appValues={appValues}
            handleValuesChange={this.handleValuesChange}
          />
        )}
        <div>
          <button className="button button-primary margin-t-big" type="submit">
            Submit
          </button>
          {goBack && (
            <button className="button" type="button" onClick={goBack}>
              Back
            </button>
          )}
        </div>
      </div>
    );
  }

  private handleChartVersionChange = (e: React.FormEvent<HTMLSelectElement>) => {
    // TODO(andres): This requires refactoring. Currently, the deploy and upgrade
    // forms behave differently. In the deployment form, a change in the version
    // changes the route but in the case of the upgrade it only changes the state
    const isUpgradeForm = !!this.props.releaseName;

    if (isUpgradeForm) {
      const { chartID, getChartVersion } = this.props;
      getChartVersion(chartID, e.currentTarget.value);
    } else {
      this.props.push(
        `/apps/ns/${this.props.namespace}/new/${this.props.chartID}/versions/${
          e.currentTarget.value
        }`,
      );
    }
  };

  private handleValuesChange = (value: string) => {
    this.props.setValues(value);
    this.props.setValuesModified();
  };

  private refreshBasicParameters = () => {
    this.setState({
      basicFormParameters: retrieveBasicFormParams(
        this.props.appValues,
        this.props.selected.schema,
      ),
    });
  };

  private renderTabs = () => {
    return (
      <div className="margin-t-normal">
        <Tabs>
          <TabList>
            <Tab onClick={this.refreshBasicParameters}>
              Basic{" "}
              <Hint reactTooltipOpts={{ delayHide: 100 }} id="basicFormHelp">
                <span>
                  This form has been automatically generated based on the chart schema.
                  <br />
                  This feature is currently in a beta state. If you find an issue please report it{" "}
                  <a target="_blank" href="https://github.com/kubeapps/kubeapps/issues/new">
                    here.
                  </a>
                </span>
              </Hint>
            </Tab>
            <Tab>Advanced</Tab>
          </TabList>
          <TabPanel>
            <BasicDeploymentForm
              params={this.state.basicFormParameters}
              handleBasicFormParamChange={this.handleBasicFormParamChange}
              appValues={this.props.appValues}
              handleValuesChange={this.handleValuesChange}
            />
          </TabPanel>
          <TabPanel>
            <AdvancedDeploymentForm
              appValues={this.props.appValues}
              handleValuesChange={this.handleValuesChange}
            />
          </TabPanel>
        </Tabs>
      </div>
    );
  };

  private handleBasicFormParamChange = (name: string, param: IBasicFormParam) => {
    return (e: React.FormEvent<HTMLInputElement>) => {
      this.props.setValuesModified();
      const value = getValueFromEvent(e);
      this.setState({
        // Change param definition
        basicFormParameters: {
          ...this.state.basicFormParameters,
          [name]: {
            ...param,
            value,
          },
        },
      });
      // Change raw values
      this.props.setValues(setValue(this.props.appValues, param.path, value));
    };
  };

  // The basic form should be rendered both if it's enabled and if there are params to show
  private shouldRenderBasicForm = () => {
    return this.props.enableBasicForm && Object.keys(this.state.basicFormParameters).length > 0;
  };
}

export default DeploymentFormBody;
