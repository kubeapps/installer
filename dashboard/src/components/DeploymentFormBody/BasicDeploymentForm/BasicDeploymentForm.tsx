import * as React from "react";
import { IBasicFormParam, IBasicFormSliderParam } from "shared/types";
import TextParam from "./TextParam";

import { getValue } from "../../../shared/schema";
import "./BasicDeploymentForm.css";
import BooleanParam from "./BooleanParam";
import SliderParam from "./SliderParam";
import Subsection from "./Subsection";

export interface IBasicDeploymentFormProps {
  params: { [name: string]: IBasicFormParam };
  handleBasicFormParamChange: (
    name: string,
    p: IBasicFormParam,
  ) => (e: React.FormEvent<HTMLInputElement>) => void;
  handleValuesChange: (value: string) => void;
  appValues: string;
}

class BasicDeploymentForm extends React.Component<IBasicDeploymentFormProps> {
  public render() {
    return (
      <div className="margin-t-normal">
        {Object.keys(this.props.params).map((paramName, i) => {
          const id = `${paramName}-${i}`;
          return (
            <div key={id}>
              {this.renderParam(
                paramName,
                this.props.params[paramName],
                id,
                this.props.handleBasicFormParamChange,
              )}
              <hr />
            </div>
          );
        })}
      </div>
    );
  }

  private isHidden = (param: IBasicFormParam) => {
    const hidden = param.hidden;
    switch (typeof hidden) {
      case "string":
        // If hidden is a string, it points to the value that should be true
        return getValue(this.props.appValues, hidden) === true;
      case "object":
        // If hidden is an object, inspect the value it points to
        return getValue(this.props.appValues, hidden.value) === hidden.condition;
      case "undefined":
        return false;
    }
  };

  private renderParam = (
    name: string,
    param: IBasicFormParam,
    id: string,
    handleBasicFormParamChange: (
      name: string,
      p: IBasicFormParam,
    ) => (e: React.FormEvent<HTMLInputElement>) => void,
  ) => {
    let paramComponent: JSX.Element = <></>;
    switch (name) {
      default:
        switch (param.type) {
          case "boolean":
            paramComponent = (
              <BooleanParam
                label={param.title || name}
                handleBasicFormParamChange={handleBasicFormParamChange}
                id={id}
                name={name}
                param={param}
              />
            );
            break;
          case "object": {
            paramComponent = (
              <Subsection
                label={param.title || name}
                handleValuesChange={this.props.handleValuesChange}
                appValues={this.props.appValues}
                renderParam={this.renderParam}
                name={name}
                param={param}
              />
            );
            break;
          }
          case "string": {
            if (param.render === "slider") {
              const p = param as IBasicFormSliderParam;
              paramComponent = (
                <SliderParam
                  label={param.title || name}
                  handleBasicFormParamChange={handleBasicFormParamChange}
                  id={id}
                  name={name}
                  param={param}
                  min={p.sliderMin || 1}
                  max={p.sliderMax || 1000}
                  unit={p.sliderUnit || ""}
                />
              );
              break;
            }
          }
          default:
            paramComponent = (
              <TextParam
                label={param.title || name}
                handleBasicFormParamChange={handleBasicFormParamChange}
                id={id}
                name={name}
                param={param}
                inputType={param.type === "integer" ? "number" : "string"}
              />
            );
        }
    }
    return (
      <div key={id} hidden={this.isHidden(param)}>
        {paramComponent}
      </div>
    );
  };
}

export default BasicDeploymentForm;
