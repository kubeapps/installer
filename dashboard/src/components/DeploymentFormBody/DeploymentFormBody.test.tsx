import { mount, shallow } from "enzyme";
import * as React from "react";

import { Tab, Tabs } from "react-tabs";
import itBehavesLike from "../../shared/specs";
import { IChartState, IChartVersion, NotFoundError } from "../../shared/types";
import { ErrorSelector } from "../ErrorAlert";
import ErrorPageHeader from "../ErrorAlert/ErrorAlertHeader";
import LoadingWrapper from "../LoadingWrapper";
import BasicDeploymentForm from "./BasicDeploymentForm/BasicDeploymentForm";
import DeploymentFormBody, {
  IDeploymentFormBodyProps,
  IDeploymentFormBodyState,
} from "./DeploymentFormBody";

const defaultProps = {
  chartID: "foo",
  chartVersion: "1.0.0",
  error: undefined,
  releaseName: undefined,
  selected: {} as IChartState["selected"],
  deployChart: jest.fn(),
  push: jest.fn(),
  fetchChartVersions: jest.fn(),
  getChartVersion: jest.fn(),
  namespace: "default",
  appValues: "foo: bar",
  valuesModified: false,
  setValues: jest.fn(),
  setValuesModified: jest.fn(),
  originalValues: undefined,
} as IDeploymentFormBodyProps;
const versions = [{ id: "foo", attributes: { version: "1.2.3" } }] as IChartVersion[];

itBehavesLike("aLoadingComponent", { component: DeploymentFormBody, props: defaultProps });

describe("renders an error", () => {
  it("renders an error if it cannot find the given chart", () => {
    const wrapper = mount(
      <DeploymentFormBody
        {...defaultProps}
        selected={{ error: new NotFoundError() } as IChartState["selected"]}
      />,
    );
    expect(wrapper.find(ErrorPageHeader).exists()).toBe(true);
    expect(wrapper.find(ErrorPageHeader).text()).toContain('Chart "foo" (1.0.0) not found');
  });

  it("renders a generic error", () => {
    const wrapper = shallow(
      <DeploymentFormBody
        {...defaultProps}
        selected={{ error: new Error() } as IChartState["selected"]}
      />,
    );
    expect(wrapper.find(ErrorSelector).exists()).toBe(true);
    expect(wrapper.find(ErrorSelector).html()).toContain("Sorry! Something went wrong");
  });
});

it("renders the full DeploymentFormBody", () => {
  const wrapper = shallow(
    <DeploymentFormBody {...defaultProps} selected={{ versions, version: versions[0] }} />,
  );
  expect(wrapper).toMatchSnapshot();
});

const initialValues = "foo: bar";
const initialSchema = { properties: { foo: { type: "string", form: true } } };
const chartVersion = {
  id: "foo",
  attributes: { version: "1.0.0", app_version: "1.0", created: "1" },
  relationships: {
    chart: {
      data: {
        name: "chart",
        description: "chart-description",
        keywords: [],
        maintainers: [],
        repo: {
          name: "repo",
          url: "http://example.com",
        },
        sources: [],
      },
    },
  },
};
const props: IDeploymentFormBodyProps = {
  ...defaultProps,
  selected: {
    ...defaultProps.selected,
    versions: [chartVersion],
    version: chartVersion,
    values: initialValues,
    schema: initialSchema,
  },
};

describe("when there are changes in the selected version", () => {
  it("initializes the local values from props when props set", () => {
    const wrapper = shallow(<DeploymentFormBody {...defaultProps} />);
    wrapper.setProps({ selected: props.selected });
    const localState: IDeploymentFormBodyState = wrapper.instance()
      .state as IDeploymentFormBodyState;
    const basicFormParameters = [
      {
        form: true,
        path: "foo",
        value: "bar",
        type: "string",
      },
    ];
    expect(localState.basicFormParameters).toEqual(basicFormParameters);
  });

  describe("when the user has not modified any value", () => {
    it("selects the original values if the version doesn't change", () => {
      const setValues = jest.fn();
      const originalValues = "foo: notBar";
      const wrapper = shallow(
        <DeploymentFormBody {...props} setValues={setValues} originalValues={originalValues} />,
      );
      wrapper.setProps({
        selected: {
          ...props.selected,
          values: "foo: ignored-value",
        },
      });
      const basicFormParameters = [
        {
          form: true,
          path: "foo",
          value: "notBar",
          type: "string",
        },
      ];
      const localState: IDeploymentFormBodyState = wrapper.instance()
        .state as IDeploymentFormBodyState;
      expect(localState.basicFormParameters).toEqual(basicFormParameters);
      expect(setValues).toHaveBeenCalledWith("foo: notBar");
    });

    it("uses the chart default values when original values are not defined", () => {
      const setValues = jest.fn();
      const wrapper = shallow(
        <DeploymentFormBody {...props} setValues={setValues} originalValues={undefined} />,
      );
      wrapper.setProps({
        selected: {
          ...props.selected,
          values: "foo: notBar",
        },
      });
      const basicFormParameters = [
        {
          form: true,
          path: "foo",
          value: "notBar",
          type: "string",
        },
      ];
      const localState: IDeploymentFormBodyState = wrapper.instance()
        .state as IDeploymentFormBodyState;
      expect(localState.basicFormParameters).toEqual(basicFormParameters);
      expect(setValues).toHaveBeenCalledWith("foo: notBar");
    });
  });

  describe("when the user has modified the values", () => {
    it("will ignore original or default values", () => {
      const setValues = jest.fn();
      const originalValues = "foo: ignored-value";
      const modifiedValues = "foo: notBar";
      const wrapper = shallow(
        <DeploymentFormBody
          {...props}
          setValues={setValues}
          originalValues={originalValues}
          valuesModified={true}
          appValues={modifiedValues}
        />,
      );
      wrapper.setProps({
        selected: {
          ...props.selected,
          values: "foo: another-ignored-value",
        },
      });
      const basicFormParameters = [
        {
          form: true,
          path: "foo",
          value: "notBar",
          type: "string",
        },
      ];
      const localState: IDeploymentFormBodyState = wrapper.instance()
        .state as IDeploymentFormBodyState;
      expect(localState.basicFormParameters).toEqual(basicFormParameters);
      expect(setValues).not.toHaveBeenCalled();
    });
  });
});

describe("when the basic form is enabled", () => {
  it("renders the different tabs", () => {
    const basicFormParameters = [
      {
        path: "wordpressUsername",
        value: "user",
      },
    ];
    const wrapper = mount(<DeploymentFormBody {...props} />);
    wrapper.setState({ appValues: "wordpressUsername: user", basicFormParameters });
    wrapper.update();
    expect(wrapper.find(LoadingWrapper)).not.toExist();
    expect(wrapper.find(Tabs)).toExist();
  });

  it("should not render the tabs if there are no basic parameters", () => {
    const wrapper = shallow(<DeploymentFormBody {...props} />);
    expect(wrapper.find(LoadingWrapper)).not.toExist();
    expect(wrapper.find(Tabs)).not.toExist();
  });

  it("changes the parameter value", () => {
    const basicFormParameters = [
      {
        path: "wordpressUsername",
        value: "user",
      },
    ];
    const setValuesModified = jest.fn();
    const setValues = jest.fn();
    const wrapper = mount(
      <DeploymentFormBody
        {...props}
        appValues="wordpressUsername: user"
        setValues={setValues}
        setValuesModified={setValuesModified}
      />,
    );
    wrapper.setState({ basicFormParameters });
    wrapper.update();

    // Fake onChange
    const input = wrapper.find(BasicDeploymentForm).find("input");
    const onChange = input.prop("onChange") as (e: React.FormEvent<HTMLInputElement>) => void;
    onChange({ currentTarget: { value: "foo" } } as React.FormEvent<HTMLInputElement>);

    expect(wrapper.state("basicFormParameters")).toEqual([
      { path: "wordpressUsername", value: "foo" },
    ]);
    expect(setValuesModified).toHaveBeenCalled();
    expect(setValues).toHaveBeenCalledWith("wordpressUsername: foo\n");
  });

  it("should update existing params if the app values change and the user clicks on the Basic tab", () => {
    const testProps = {
      ...props,
      selected: {
        ...props.selected,
        schema: { properties: { wordpressUsername: { type: "string", form: true } } },
      },
    };
    const basicFormParameters = [
      {
        path: "wordpressUsername",
        value: "user",
      },
    ];
    const wrapper = mount(<DeploymentFormBody {...testProps} />);
    wrapper.setState({ basicFormParameters });
    wrapper.setProps({ appValues: "wordpressUsername: foo" });
    wrapper.update();

    const tab = wrapper
      .find(Tab)
      .findWhere(t => !!t.text().match("Basic"))
      .first();
    tab.simulate("click");

    expect(wrapper.state("basicFormParameters")).toMatchObject([
      {
        path: "wordpressUsername",
        value: "foo",
      },
    ]);
  });

  it("handles a parameter as a number", () => {
    const setValues = jest.fn();
    const basicFormParameters = [
      {
        path: "replicas",
        value: 1,
        type: "integer",
      },
    ];
    const wrapper = mount(<DeploymentFormBody {...props} setValues={setValues} />);
    wrapper.setState({ basicFormParameters });
    wrapper.setProps({ appValues: "replicas: 1" });
    wrapper.update();

    // Fake onChange
    const input = wrapper.find(BasicDeploymentForm).find("input");
    const onChange = input.prop("onChange") as (e: React.FormEvent<HTMLInputElement>) => void;
    onChange({ currentTarget: { value: "2", valueAsNumber: 2, type: "number" } } as React.FormEvent<
      HTMLInputElement
    >);

    expect(wrapper.state("basicFormParameters")).toEqual([
      {
        path: "replicas",
        value: 2,
        type: "integer",
      },
    ]);
    expect(setValues).toHaveBeenCalledWith("replicas: 2\n");
  });

  it("handles a parameter as a boolean", () => {
    const setValues = jest.fn();
    const basicFormParameters = [
      {
        path: "enableMetrics",
        value: false,
        type: "boolean",
      },
    ];
    const wrapper = mount(<DeploymentFormBody {...props} setValues={setValues} />);
    wrapper.setState({ basicFormParameters });
    wrapper.setProps({ appValues: "enableMetrics: false" });
    wrapper.update();

    // Fake onChange
    const input = wrapper.find(BasicDeploymentForm).find("input");
    const onChange = input.prop("onChange") as (e: React.FormEvent<HTMLInputElement>) => void;
    onChange({
      currentTarget: { value: "true", checked: true, type: "checkbox" },
    } as React.FormEvent<HTMLInputElement>);

    expect(wrapper.state("basicFormParameters")).toEqual([
      {
        path: "enableMetrics",
        value: true,
        type: "boolean",
      },
    ]);
    expect(setValues).toHaveBeenCalledWith("enableMetrics: true\n");
  });
});

it("goes back when clicking in the Back button", () => {
  const goBack = jest.fn();
  const wrapper = shallow(<DeploymentFormBody {...props} goBack={goBack} />);
  const backButton = wrapper.find(".button").filterWhere(i => i.text() === "Back");
  expect(backButton).toExist();
  // Avoid empty or submit type
  expect(backButton.prop("type")).toBe("button");
  backButton.simulate("click");
  expect(goBack).toBeCalled();
});
