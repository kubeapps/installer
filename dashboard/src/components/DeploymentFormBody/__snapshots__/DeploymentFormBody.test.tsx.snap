// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loading spinner matches the snapshot 1`] = `
<LoadingWrapper
  loaded={false}
  type={0}
/>
`;

exports[`renders the full DeploymentFormBody 1`] = `
<div>
  <div>
    <label
      htmlFor="chartVersion"
    >
      Version
    </label>
    <select
      id="chartVersion"
      onChange={[Function]}
      required={true}
      value="1.2.3"
    >
      <option
        key="foo"
        value="1.2.3"
      >
        1.2.3
         
      </option>
    </select>
  </div>
  <AdvancedDeploymentForm
    appValues="foo: bar"
    handleValuesChange={[Function]}
  />
  <div
    className="margin-t-big"
  >
    <button
      className="button button-primary"
      type="submit"
    >
      Submit
    </button>
  </div>
</div>
`;
